lib/ansible/galaxy/dependency_resolution/providers.py

```diff
@@ -31,8 +32,33 @@
 from ansible.module_utils.six import string_types
 from ansible.utils.version import SemanticVersion
 
+from collections.abc import Set
 from resolvelib import AbstractProvider
 
+PINNED_CANDIDATE_ATTRS = ('fqcn', 'ver', 'src', 'type')
+
+
+class PinnedCandidateRequests(Set):
+    def __init__(self, candidates):
+        self.candidates = set(candidates)
+
+    def __iter__(self):
+        return iter(self.candidates)
+
+    def __contains__(self, value):
+        for candidate in self.candidates:
+            for attr in PINNED_CANDIDATE_ATTRS:
```

Comment:
Looks like this is just
```suggestion
        for candidate, attr in itertools.product(self.candidates, PINNED_CANDIDATE_ATTRS):
```

Reply:
Consolidating into one loop breaks what this is doing. Consider:
```
for candidate, attr in itertools.product(candidates, ('fqcn', 'ver', 'src', 'type')):
    if getattr(value, attr) != getattr(candidate, attr):
        break
else:
    return True
```
If candidates is an empty list, it will return True. If the list is not empty and contains a non-matching candidate before a matching one, it would never return True. The inner loop is a `for-else` that should occur per candidate.